manpages/mrctaper.man
.na
.nh
.TH mrctaper 1 2.42 BL3DEMC
.SH NAME
mrctaper \- Tapers edges of images between actual and filled-in values
.SH SYNOPSIS
mrctaper  [-i]  [-t #]  [-z min,max]  input_file  [output_file]
.SH DESCRIPTION
mrctaper will taper the edges between actual image data and areas that have
been filled in with a constant value, thus eliminating the sharp transitions
produced when images are transformed
by programs such as newstack(1) or blendmont(1).  By default, the program
will taper outside the actual image data by replicating the pixels along the 
edge and attenuating their intensity toward the filled-in value for pixels 
farther from the edge.  This makes the features
on the edge of the area appear smeared out into the filled-in area.  This
tapering will preserve all of the actual data but may produce undesirable 
artifacts in some cases.  Optionally, the program can taper inside, by 
attenuating the intensity of actual pixels toward the filled-in value and 
doing so progressively less for pixels farther from the edge.  This method
will throw away some information but will be most successful at preventing 
artifacts in Fourier transforms.

.SH OPTIONS
.TP
.B -i
Taper inside rather than outside.
.TP
.B -t \fI#\fR
Taper over the given number of pixels.  The default value is 1% of the mean
of the X and Y sizes of the image, but at least 16 pixels and no more than 127.
.TP
.B -z \fImin,max\fR
Operate only on sections from
.B min
to
.B max.
Sections are numbered from 0.  If a new output file is being
created, it will contain only the sections from min to max in the input file.
.SH FILES
If no output file name is given, the data are written back into the input file.
.SH AUTHOR
David Mastronarde  <mast@colorado.edu>
.SH SEE ALSO
newstack(1), blendmont(1)


History
IMOD History
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
12/3/13: In eTomo - PEET, excluding the quoted values functionality.
(Bug# 1631)

11/30/13: Changes to Tiltalign to deal with patch tracking data: it can select
a subset of patches for the global alignment to avoid solving for too many
variables; it can do robust fitting with weights assigned either to individual
points or to whole contours.

11/28/13: Made changes in the search routine used by Tiltalign/Beadtrack to
reduce the high risk of "minimization errors" when starting close to the final
solution; also ignore particular errors after first round of robust fitting.

11/28/13: In eTomo allowing quoted values in autodocs.(Bug# 1631)

11/21/13: IMOD_4-6-33
11/21/13: Changed Corrsearch3d to use dynamic allocation, load from file more 
efficiently, and use multiple threads for some steps.

11/20/13: In eTomo - Final Aligned Stack, prevented the creation of a bad
rotation.xf file when the final aligned image stack is not present.(Bug# 1762)

11/20/13: In eTomo - Serial Sections, fixed the Newstack progress bar.
(Bug# 1749)

11/20/13: Fixed 3dmod crash from loading many files through startup window.

11/20/13: Made multi-line INFO and WARNING messages more parsable in
etomo_err.log.(Bug# 1694)

11/19/13: In eTomo - Combine, prevented the dataset from being damaged when
the Combine dialog is opened too early.(Bug# 1652)

11/19/13: In uitest, fixed a problem with finding the main panel.(Bug# 1761)

11/18/13: In eTomo made multi-line error messages parsable by adding an empty
line after each one.(Bug# 1694)

11/18/13: In eTomo - Coarse Alignment, added support for tiltxcorr parameters
SearchMagChanges and ViewsWithMagChanges.(Bug# 1760)

11/15/13: In eTomo added support for parallel processing using multiple GPUs
per computer.(Bug# 1436)

11/15/13: In eTomo reduced the excessive debug output.(Bug# 1763)

11/13/13: In eTomo - Serial Sections, added a Sobel filter checkbox and
improved the behavior of the alignment panel.(Bug# 1755)

11/11/13: IMOD_4-6-32
11/5/13: Fixed bug that crashed Imodfindbeads when adjusting size downward.

11/4/13: In eTomo - Fiduacial Model dialog, fixed a problem with deleting
autofidseed temporary files in Windows.(Bug# 1759)

11/4/13: Make xfalign apply shift limits in initial cross-correlation.

11/4/13: Fixed scale bar drawing from montage snapshot of model view window.

11/4/13: Made mrc2tif use 4 or 5 digits for all output file numbers if needed.

11/2/13: Switched display of residual model to show arrows.

11/1/13: In eTomo - Serial Sections, improved the context sensitive pop-up
menus.(Bug# 1748)

11/1/13: In eTomo - Fiducial Model, added fields for AdjustSizes and
ClusteredPointsAllowed to the Autofidseed panel.(Bug# 1743)

11/1/13: Made Tiltxcorr able to find mag changes at selected views; for a 
bidirectional series, Copytomocoms puts the option with the view number in
xcorr.com.

11/1/13: Fixed bug in imodtrans -3 option introduced 12/8/12.

10/30/13: IMOD_4-6-31
10/29/13: Fixed a problem in eTomo where the parallel panel did not appear
when eTomo reconnected to a parallel process.  This problem appeared when
Pause and Resume had been previously used.(Bug# 1757)

10/29/13: Fixed the eTomo version string.(Bug# 1744)

10/28/13: In eTomo fixed a bug where stack header data was not getting into
the dataset file when eTomo was run from batchruntomo.(Bug# 1758)

10/27/13: Added option to Tilt to get output in parallel (X/Y) slices without
inverting handedness. 

10/25/13: In eTomo removed dependencies on Cygwin.(Bug# 1713)

10/25/13: Changes in Tiltxcorr: shift limits are now applied in an elliptical
rather than rectangular region; implemented a new, more selective method for
excluding the central peak due to fixed pattern noise; added an option to
compute normalized correlation coefficients at multiple peak positions and
pick the highest (not useful so far).

10/25/13: Changed Beadtrack's elongation file output so that automatic seed
picking would not fail with only 3 beads.

10/24/13: In eTomo - Combine, removed the message complaining about tomograms
A and B not being the same size in X and Y, as this is not an error.
(Bug# 1651)

10/24/13: In eTomo, preventing embedded spaces in dataset names.(Bug# 1667)

10/24/13: Added a antialiasFilter field for the newstack panels in eTomo -
Coarse Alignment and Final Aligned Stack.(Bug# 1737)

10/23/13: In eTomo resizing when the Cleanup dialog is finished.(Bug# 1636)

10/22/13: Fixed another warping-related crash in Midas.

10/22/13: Made it possible to start Zap rubber band from ANY corner.

10/21/13: Fixed a problem that crashed 3dmod when a slicer angle is very
small, such as when doing mouse-rotating.

10/18/13: Added toolbar buttons to Zap and Slicer for drawing a single thick
arrow for snapshots; also added an object property to draw arrowheads at the
ends of contours.

10/17/13: In eTomo - Clean up, added unused, optional image files (such as the
output from mtffilter) to the file chooser.(Bug# 1509)

10/17/13: In eTomo - Clean up, added SIRT .vsr## output files.(Bug# 1533)

10/15/13: Fixed numerical problems in drawing code that crashed Midas on big
images in 64-bit Windows.

10/14/13: Added GiantCriterion, BigDiffCriterion, and ExtraLargeRadius to
eTomo - Preprocessing.(Bug# 1736)

10/9/13: Fixed a problem with running tomodataplots from eTomo under Java 1.7
that prevented the graphs from coming up.(Bug# 1754)

10/9/13: In eTomo fixed a parallel processing problem which prevented the
Resume button from working when the process was pausing.  This problem
appeared in only in reconnected processes.(Bug# 1729)

10/9/13: In eTomo - Nonlinear Anisotropic Diffusion, some test.K... file names
will differ in different Java versions.  This bug will not be fixed because it
does not appear to cause any harm.  Only the automated tests where modified.
(Bug# 1753)


mrc/mrctaper.c
/*  
 *  mrctaper.c -- tapers edges of images where they have been filled
 *
 *  Author: David Mastronarde   email: mast@colorado.edu
 *
 *  Copyright (C) 1995-2005 by Boulder Laboratory for 3-Dimensional Electron
 *  Microscopy of Cells ("BL3DEMC") and the Regents of the University of 
 *  Colorado.  See dist/COPYRIGHT for full copyright notice.
 *
 *  $Id$
 */

#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include "mrcslice.h"
#include "mrcfiles.h"
#include "b3dutil.h"
#include "sliceproc.h"

#define DEFAULT_TAPER 16

void mrctaper_help(char *name)
{
  fprintf(stderr,"Usage: %s [-i] [-t #] [-z min,max] input_file [output_file]"
          "\n", name);
  puts("Options:");
  puts("\t-i\tTaper inside (default is outside).");
  printf("\t-t #\tTaper over the given # of pixels (default %d or 1%% of size)"
         ".\n", DEFAULT_TAPER);
  puts("\t-z min,max\tDo only sections between min and max.");
  puts("\tWith no output file, images are written back to input file.");
}

int main( int argc, char *argv[] )
{

  int    i = 0;
  FILE   *fin, *fout;
  struct MRCheader hdata, hout;
  struct MRCheader *hptr;
  unsigned char *buf;
  int bsize, csize, dsize;
  int inside = 0;
  int ntaper = DEFAULT_TAPER;
  int taperEntered = 0;
  Islice slice;
  int zmin = -1, zmax = -1;
  int secofs;
  char *progname = imodProgName(argv[0]);

  if (argc < 2){
    fprintf(stderr, 
            "%s version %s\n", progname, VERSION_NAME);
    imodCopyright();
    mrctaper_help(progname);
    exit(3);
  }

  for (i = 1; i < argc; i++)
    if (argv[i][0] == '-')
      switch (argv[i][1]){
          
      case 'i':
        inside = 1;
        break;

      case 't':
        taperEntered = 1;
        if (argv[i][2] != 0x00)
          sscanf(argv[i], "-t%d", &ntaper);
        else
          sscanf(argv[++i], "%d", &ntaper);
        break;

      case 'z':
        if (argv[i][2] != 0x00)
          sscanf(argv[i], "-z%d%*c%d", &zmin, &zmax);
        else
          sscanf(argv[++i], "%d%*c%d", &zmin, &zmax);
        break;

      default:
        printf("ERROR: %s - illegal option\n", progname);
        mrctaper_help(progname);
        exit(1);
        break;
            
      }else break;

     

  if (i < (argc - 2) || i == argc){
    mrctaper_help(progname);
    exit(3);      
  }

  if (ntaper < 1 || ntaper > 127) {
    printf("ERROR: %s - Taper must be between 1 and 127.\n",
            progname);
    exit(3);
  }

  if (i < argc - 1)
    fin = fopen(argv[i++], "rb");
  else
    fin = fopen(argv[i++], "rb+");

  if (fin == NULL){
    printf("ERROR: %s - Opening %s.\n", progname, argv[i - 1]);
    exit(3);
  }
  if (mrc_head_read(fin, &hdata)) {
    printf("ERROR: %s - Can't Read Input File Header.\n", progname);
    exit(3);
  }

  if (sliceModeIfReal(hdata.mode) < 0) {
    printf("ERROR: %s - Can operate only on byte, integer and "
            "real data.\n", progname);
    exit(3);
  }

  if (!taperEntered) {
    ntaper = (hdata.nx + hdata.ny) / 200;
    ntaper = B3DMIN(127, B3DMAX(DEFAULT_TAPER, ntaper));
    printf("Tapering over %d pixels\n", ntaper);
  }
     
  if (zmin == -1 && zmax == -1) {
    zmin = 0;
    zmax = hdata.nz - 1;
  } else {
    if (zmin < 0)
      zmin = 0;
    if (zmax >= hdata.nz)
      zmax = hdata.nz - 1;
  }
     
  if (i < argc){
    fout = fopen(argv[i], "wb");
    if (fout == NULL) {
      printf("ERROR: %s - Opening %s.\n", progname, argv[i]);
      exit(3);
    }
    hout = hdata;
    /* DNM: eliminate extra header info in the output, and mark it as not swapped  */
    mrcInitOutputHeader(&hout);
    hptr = &hout;
    hout.nz = zmax + 1 - zmin;
    hout.mz = hout.nz;
    hout.zlen = hout.nz;
    secofs = zmin;
  }else{
    /* DNM 6/26/02: it it OK now */
    /* if (hdata.swapped) {
       fprintf(stderr, "%s: Cannot write to byte-swapped file.\n", progname);
       exit(3);
       } */
    hptr = &hdata;
    fout = fin;
    secofs = 0;
  }
     
  mrc_getdcsize(hdata.mode, &dsize, &csize);

  bsize = hdata.nx * hdata.ny;
  buf = (unsigned char *)malloc(dsize * csize * bsize);
     
  if (!buf){
    printf("ERROR: %s - Couldn't get memory for slice.\n", progname);
    exit(3);
  }
  sliceInit(&slice, hdata.nx, hdata.ny, hdata.mode, buf);

  for (i = zmin; i <= zmax; i++) {
    printf("\rDoing section #%4d", i);
    fflush(stdout);
    if (mrc_read_slice(buf, fin, &hdata, i, 'Z')) {
      printf("\nERROR: %s - Reading section %d.\n", progname, i);
      exit(3);
    }
      
    if (sliceTaperAtFill(&slice, ntaper, inside)) {
      printf("\nERROR: %s - Can't get memory for taper operation.\n",
              progname);
      exit(3);
    }
          
    if (mrc_write_slice(buf, fout, hptr, i - secofs, 'Z')) {
      printf("\nERROR: %s - Writing section %d.\n", progname, i);
      exit(3);
    }
  }
  puts("\nDone!");

  mrc_head_label(hptr, "mrctaper: Image tapered down to fill value at "
                 "edges");

  mrc_head_write(fout, hptr);

  return(0);
}
